src/backend/distributed/utils/shard_pruning.c

```diff
@@ -0,0 +1,1239 @@
+											cacheEntry->hasUniformHashDistribution);
+
+		if (shardIndex == INVALID_SHARD_INDEX)
+		{
+			return NIL;
+		}
+		else if (shardInterval &&
+				 sortedShardIntervalArray[shardIndex]->shardId != shardInterval->shardId)
+		{
+			return NIL;
+		}
+		else
+		{
+			return list_make1(sortedShardIntervalArray[shardIndex]);
+		}
+	}
+
+	/*
+	 * FIXME: We could perform ExhaustivePrune() on the individual
+	 * shardInterval, in case there's both an = and a < / >? It'd probably be
+	 * better to pre-compute that by checking a PruningInstance's
+	 * less[Equal]Consts/greater[Equal]Consts against equalConsts.
+	 */
+	if (shardInterval)
+	{
+		return list_make1(shardInterval);
+	}
+
+	/*
+	 * Should never get here for hashing, either we've filtered down to zero
+	 * or one shard, and returned.
+	 */
+	Assert(context->partitionMethod != DISTRIBUTE_BY_HASH);
+
+	/*
+	 * Next method: binary search with fuzzy boundaries. Can't trivially do so
+	 * if shards have overlapping boundaries.
+	 *
+	 * TODO: If we kept shard intervals separately sorted by both upper and
+	 * lower boundaries, this should be possible?
+	 */
+	if (!cacheEntry->hasOverlappingShardInterval && (
+			prune->greaterConsts || prune->greaterEqualConsts ||
+			prune->lessConsts || prune->lessEqualConsts))
+	{
+		return PruneWithBoundaries(cacheEntry, context, prune);
+	}
+
+	/*
+	 * Brute force: Check each shard.
+	 */
+	return ExhaustivePrune(cacheEntry, context, prune);
+}
+
+
+/*
+ * PerformCompare invokes comparator with prepared values, check for
+ * unexpected NULL returns.
+ */
+static int
+PerformCompare(FunctionCallInfoData *compareFunctionCall)
+{
+	Datum result = FunctionCallInvoke(compareFunctionCall);
+
+	if (compareFunctionCall->isnull)
+	{
+		elog(ERROR, "function %u returned NULL", compareFunctionCall->flinfo->fn_oid);
+	}
+
+	return DatumGetInt32(result);
+}
+
+
+/*
+ * PerformValueCompare invokes comparator with a/b, and checks for unexpected
+ * NULL returns.
+ */
+static int
+PerformValueCompare(FunctionCallInfoData *compareFunctionCall, Datum a, Datum b)
+{
+	compareFunctionCall->arg[0] = a;
+	compareFunctionCall->argnull[0] = false;
+	compareFunctionCall->arg[1] = b;
+	compareFunctionCall->argnull[1] = false;
+
+	return PerformCompare(compareFunctionCall);
+}
+
+
+/*
+ * LowerShardBoundary returns the index of the first ShardInterval that's >=
+ * (if includeMax) or > partitionColumnValue.
+ */
+static int
+LowerShardBoundary(Datum partitionColumnValue, ShardInterval **shardIntervalCache,
+				   int shardCount, FunctionCallInfoData *compareFunction, bool includeMax)
+{
+	int lowerBoundIndex = 0;
+	int upperBoundIndex = shardCount;
+
+	Assert(shardCount != 0);
+
+	/* setup partitionColumnValue argument once */
+	compareFunction->arg[0] = partitionColumnValue;
+	compareFunction->argnull[0] = false;
+
+	while (lowerBoundIndex < upperBoundIndex)
+	{
+		int middleIndex = (lowerBoundIndex + upperBoundIndex) / 2; /* XXX: overflow? */
```

Comment:
Why?  Practically it'll currently not because you'll grow very bored before creating close to enough shards, but if upperBoundIndex is something like 2**30, and lowerBoundIndex is increased to 2**29 by one probe, you'll get an overflow, no?